<?xml version="1.0"?>

<launch>
    <arg name="use_rviz" default="false" />
    <arg name="controller" default="PD" doc="opt: mpc, L1,PD" />
    <!-- for amcl -->    
    <arg name="init_x" default="0.0" />
    <arg name="init_y" default="0.0" />
    <arg name="init_a" default="0.0" />

    <!-- Map server -->
    <node name="map_server" pkg="map_server" type="map_server" args="$(find art_racecar)/map/mymap.yaml"/>


    <!-- ODOMETRY -->
    <!--rf2o_Laser_Odometry-->
     <include file="$(find art_racecar)/launch/includes/rf2o.launch.xml" />

    <!--wheel_encoder_odometry-->
     <!--include file="$(find art_racecar)/launch/includes/wheel_odometry.launch.xml" /-->

    <!-- Robot_Localization -->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_se" clear_params="true">
        <rosparam command="load" file="$(find art_racecar)/param/ekf_params.yaml" />
    </node>

    <!-- obstacle -->
    <!--<node pkg="laser_judge_obstacle" type="laser_judge_obstacle" name="laser_judge_obstacle">
    </node>-->

    <!-- Localization -->
    <!-- AMCL -->
    <include file="$(find art_racecar)/launch/includes/amcl.launch.xml">
        <arg name="init_x" value="$(arg init_x)"/>
        <arg name="init_y" value="$(arg init_y)"/>
        <arg name="init_a" value="$(arg init_a)"/>
    </include>

    <!-- Navstack -->
    <node pkg="move_base" type="move_base" respawn="false" name="move_base">
        <!-- local planner -->
       
	<!-- <param name="base_global_planner" value="navfn/NavfnROS"/> -->
      
        <param name="base_global_planner" value="global_planner/GlobalPlanner"/>  
        <!--<param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS"/>-->  
        <param name="base_local_planner" value="teb_local_planner/TebLocalPlannerROS"/>

        <rosparam file="$(find art_racecar)/param/teb_local_planner_params.yaml" command="load"/>
        <!-- costmap layers -->    
        <rosparam file="$(find art_racecar)/param/local_costmap_params.yaml" command="load"/>
        <rosparam file="$(find art_racecar)/param/global_costmap_params.yaml" command="load"/> 
        <!-- move_base params -->
        <rosparam file="$(find art_racecar)/param/base_global_planner_params.yaml" command="load"/>
        <rosparam file="$(find art_racecar)/param/move_base_params.yaml" command="load"/>
        <remap from="/odom" to="/odometry/filtered"/>
	<remap from="/cmd_vel" to="/fake_vel"/>     
    </node>
   <!--<node name="laserCorrection" pkg="laser_correction"  type="laserCorrection" output="screen" >
	<remap from="/odom" to="/odometry/filtered"/>
    </node>-->
    <!-- L1 controller -->  
    <node pkg="art_racecar" type="art_car_controller" respawn="false" name="art_car_controller" output="screen" if="$(eval controller == 'L1')">
	<rosparam file="$(find art_racecar)/param/L1/L1_params.yaml" command="load" />
	<rosparam file="$(find art_racecar)/param/L1/ADRC_params.yaml" command="load" />
        <!-- remap from="/odometry/filtered" to="odom" / --> 
        <!--<remap from="/move_base_node/NavfnROS/plan" to="/move_base/GlobalPlanner/plan" />-->
    </node>
    <!--  ****** MPC Node ******  -->
    <!--<node name="MPC_Node" pkg="art_racecar" type="MPC_Node" output="screen" if="$(eval controller == 'mpc')" >
        <rosparam file="$(find art_racecar)/param/mpc/mpc_params.yaml" command="load" />
        <rosparam file="$(find art_racecar)/param/mpc/racecar_params.yaml" command="load" />
    </node>-->
    <node name="PD_controller" pkg="art_racecar" type="PD_controller" output="screen" if="$(eval controller == 'PD')" >
	<rosparam file="$(find art_racecar)/param/PD/PD_params.yaml" command="load" />
	</node>
    <!--<node pkg="simple_navigation_goals" type="simple_navigation_goals"  name="simple_navigation_goals">
        <param name="goal_fist_x" value="8.0"/> 
        <param name="goal_fist_y" value="0.0"/>
        <param name="goal_second_x" value="15.5"/>
        <param name="goal_second_y" value="0.0"/>
    </node>-->
    <!-- Rviz -->
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find art_racecar)/rviz/amcl.rviz" if="$(arg use_rviz)" />
</launch>
